evm_loader/program/src/executor/state.rs

```diff
@@ -285,17 +285,39 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
     }
 
     async fn code_hash(&self, from_address: Address, chain_id: u64) -> Result<[u8; 32]> {
-        use solana_program::keccak::hash;
-
-        for action in &self.actions {
-            if let Action::EvmSetCode { address, code, .. } = action {
-                if &from_address == address {
-                    return Ok(hash(code).to_bytes());
-                }
-            }
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // https://eips.ethereum.org/EIPS/eip-161
+
+        #[maybe_async]
+        async fn data_account_exists<B: AccountStorage>(
+            state: &ExecutorState<'_, B>,
+            address: Address,
+            chain_id: u64,
+        ) -> Result<bool> {
+            Ok(state.nonce(address, chain_id).await? > 0
+                || state.balance(address, chain_id).await? > 0)
         }
 
-        Ok(self.backend.code_hash(from_address, chain_id).await)
+        // FIXME: Can we modify self.code to return Option<Buffer> or Option<&[u8]>?
+
+        // FIXME: Because buffer is returned by value, we need to store buffer in order to store a
+        // reference to the bytes. I could use Option<Buffer> instead, but that is storing a more
+        // powerful type than I need. I just need a byte slice.
+        let buffer = self.code(from_address).await?;
+        let bytes_to_hash: Option<&[u8]> = if !buffer.buffer_is_empty() {
+            // A program account exists at the address.
+            Some(&buffer)
+        } else if data_account_exists(self, from_address, chain_id).await? {
+            // A data account exists at the address.
+            Some(&[])
+        } else {
+            // No account exists at the address.
+            None
+        };
+
+        Ok(bytes_to_hash
+            .map(|bytes| solana_program::keccak::hash(bytes).to_bytes())
+            .unwrap_or_default())
```

Comment:
I suggest using early returns in order to make control flow more explicit. Something like:
```rust
        // https://eips.ethereum.org/EIPS/eip-1052
        // https://eips.ethereum.org/EIPS/eip-161

        let code = self.code(address).await?;

        if !code.is_empty() {
            return Ok(keccak::hash(&code).to_bytes());
        }

        if self.nonce(address, chain_id).await? == 0 && self.balance(address, chain_id).await? == 0
        {
            return Ok(<[u8; 32]>::default());
        }

        // equal to "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
        return Ok(keccak::hash(&[]).to_bytes());
```

Reply:
In my opinion, the proposed implementation is clearer. It is easy to see that we always produce a hash from some bytes. It is also easy to see how those bytes are produced. The control flow is clearer than when there are multiple return statements.